tests/unit/recommenders/evaluation/test_python_time.py

```diff
@@ -0,0 +1,257 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import numpy as np
+import pandas as pd
+import pytest
+from unittest.mock import Mock
+from sklearn.preprocessing import minmax_scale
+
+from recommenders.utils.constants import (
+    DEFAULT_USER_COL,
+    DEFAULT_ITEM_COL,
+    DEFAULT_RATING_COL,
+    DEFAULT_PREDICTION_COL,
+    SEED,
+)
+from recommenders.evaluation.python_evaluation import (
+    merge_rating_true_pred,
+    merge_ranking_true_pred,
+    rmse,
+    mae,
+    rsquared,
+    exp_var,
+    get_top_k_items,
+    precision_at_k,
+    recall_at_k,
+    ndcg_at_k,
+    map_at_k,
+    auc,
+    logloss,
+)
+import random
+from recommenders.utils.timer import Timer
+
+random.seed(SEED)
+DATA_USER_NUM = 10000
+DATA_ITEM_NUM = DATA_USER_NUM * 2
+DATA_SAMPLE_NUM = DATA_USER_NUM * 1000
+DATA_RATING_MAX = 5
+
+# fmt: off
+DF_RATING_TRUE = pd.DataFrame(
+    {
+        DEFAULT_USER_COL: random.choices(range(0, DATA_USER_NUM), k=DATA_SAMPLE_NUM),
+        DEFAULT_ITEM_COL: random.choices(range(0, DATA_ITEM_NUM), k=DATA_SAMPLE_NUM),
+        DEFAULT_RATING_COL: random.choices(range(1, DATA_RATING_MAX+1), k=DATA_SAMPLE_NUM),
+    }
+)
+
+
+DF_RATING_PRED = pd.DataFrame(
+    {
+        DEFAULT_USER_COL: random.choices(range(0, DATA_USER_NUM), k=DATA_SAMPLE_NUM),
+        DEFAULT_ITEM_COL: random.choices(range(0, DATA_ITEM_NUM), k=DATA_SAMPLE_NUM),
+        DEFAULT_PREDICTION_COL: random.choices(range(1, DATA_RATING_MAX+1), k=DATA_SAMPLE_NUM),
+        # DEFAULT_RATING_COL: random.choices(range(1, RATING_MAX+1), k=SAMPLE_NUM),
+    }
+)
+
+
+DF_RATING_NOHIT = pd.DataFrame(
+    {
+        DEFAULT_USER_COL: random.choices(range(0, DATA_USER_NUM), k=DATA_SAMPLE_NUM),
+        DEFAULT_ITEM_COL: [DATA_ITEM_NUM] * DATA_SAMPLE_NUM,
+        DEFAULT_PREDICTION_COL: random.choices(range(1, DATA_RATING_MAX+1), k=DATA_SAMPLE_NUM),
+    }
+)
+
+
+@pytest.fixture
+def rating_true():
+    return DF_RATING_TRUE.copy()
+
+
+@pytest.fixture
+def rating_pred():
+    return DF_RATING_PRED.copy()
+
+
+@pytest.fixture
+def rating_nohit():
+    return DF_RATING_NOHIT.copy()
+# fmt: on
+
+
+@pytest.fixture
+def rating_true_binary(rating_true):
+    # Convert true ratings to binary
+    rating_true[DEFAULT_RATING_COL] = rating_true[DEFAULT_RATING_COL].apply(
+        lambda x: 1.0 if x >= 3 else 0.0
+    )
+    return rating_true
+
+
+@pytest.fixture
+def rating_pred_binary(rating_pred):
+    # Normalize the predictions
+    rating_pred[DEFAULT_PREDICTION_COL] = minmax_scale(
+        rating_pred[DEFAULT_PREDICTION_COL].astype(float)
+    )
+    return rating_pred
+
+
+# The following time thresholds are benchmarked on Azure
+# Standard_D14_v2 VM, with Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz,
+# 16 cores, 112 GB RAM, 800 GB disk.
+# The thresholds are calculated by MEAN + 5 * STANDARD DEVIATION.
```

Comment:
Is it possible to skip the tests if the CPU were inferior to the one mentioned here?

Reply:
> Is it possible to skip the tests if the CPU were inferior to the one mentioned here?

Miguel and I had a discussion about this. We were thinking about adding a CPU benchmark function like is_cpu_powerful at the beginning of the module to test the power of the CPU. The function should return true if the target CPU finishes the task faster than our one. Then we add skipif using is_cpu_powerful as the condition to all other time tests.

What do you think?

Reply:
One clarification Simon, the skipif clause is the same as [here](https://github.com/microsoft/recommenders/blob/main/tests/unit/recommenders/tuning/test_nni_utils.py#L55).

Something that would be good to clarify Chuyang is how do you propose to implement the function?

Reply:
I guess a naive way to do this, is to 1) create the synthetic datasets rating_true and rating_pred of the same size, and 2) run map_at_k(). I choose map_at_k() here because its implementation contains the merging function and get_top_k.
We then take the runtime of the two steps combined as the criterion, to decide if the target CPU is more powerful than ours. If it is faster than our baseline, the function returns true and the tests can proceed.

Reply:
Sounds good.

Reply:
This is a very interesting topic. We should think of a general way of approaching this, for example, let's say we also want to use this approach to get the time performance of other tests, like the splitters, we would probably want to have a function that measures the power of the CPU that is agnostic to the specific function we are testing. Maybe a function that breaks a hash? how do people measure the performance of CPUs?

@ChuyangKe if you create an issue in the repo explaining the function that we want to want to create, why we need it, and where we are going to use it, then I can call the community to get feedback and ideas about this.

@simonzhaoms are you ok with leaving the cpu function out of this PR and merge this one once it is finished?